src/AsyncFunction.Wrap.tt

```diff
@@ -0,0 +1,143 @@
+        /// </summary>
+        /// <returns>A <c><#= returnTypeCount > 0 ? "Func" : "Action" #></c> to invoke this function, or <c>null</c> if the type signature is incompatible.</returns>
+        public <#= delegateType #>? Wrap<#= returnTypeCount > 0 ? "Func" : "Action" #><#= methodGenerics #>()
+        {
+            if (store is null || IsNull)
+            {
+                throw new InvalidOperationException("Cannot wrap a null function reference.");
+            }
+
+            // Try to retrieve it from the cache. if it's cached it must have passed the type
+            // signature check already, so it's safe to do this before CheckTypeSignature.
+            if (_wrapperCache?.GetType() == typeof(<#= delegateType #>))
+            {
+                return (<#= delegateType #>)_wrapperCache;
+            }
+
+            // Check that the requested type signature is compatible
+            <# if (callbackParameterTypeExpressions.Length == 0) { #>
+var parameterTypes = Array.Empty<Type>();
+            <# } else { #>
+var parameterTypes = new Type[] { <#= callbackParameterTypeExpressions #>};
+            <# } #>
+
+            // Check if the requested type signature is compatible with the function being wrapped
+            if (!CheckTypeSignature(<#= callbackReturnTypeExpression #>, parameterTypes))
+            {
+                return null;
+            }
+
+            // Fetch a converter for each parameter type to box it
+<#= parameterConverters #>
+<#
+    if (returnTypeCount > 0)
+    { 
+#>
+            // Create a factory for the return type
+            var factory = ReturnTypeFactory<TResult>.Create();
+
+<# 
+    } 
+#>
+            // Determine how much space to allocate for params/results
+            var allocCount = <#= parameterCount.ToString(CultureInfo.InvariantCulture) #> + Results.Count;
+
+            <#= delegateType #> result = (<#
+    for (int x = 0; x < parameterCount; x++)
+    {
+        if (x >= 1)
+        {
+            #>, <#
+        }
+
+        #>p<#= x.ToString(CultureInfo.InvariantCulture) #><#
+
+    }
+#>) =>
+            {
+                // Allocate space for both the arguments and the results.
+                var argsAndResults = ArrayPool<Value>.Shared.Rent(allocCount);
+                try
+                {
+<#
+    for (int x = 0; x < parameterCount; x++)
+    {
+        string xStr = x.ToString(CultureInfo.InvariantCulture);
+#>
+                    argsAndResults[<#= xStr #>] = Value.FromValueBox(
+                        convT<#= parameterCount is 1 ? "" : (x + 1).ToString(CultureInfo.InvariantCulture) #>.Box(p<#= xStr #>)
+                    );
+<#
+    }
+
+    if (returnTypeCount > 0)
+    {
+#>
+                    return InvokeWithReturn(argsAndResults, factory);
+<#
+    }
+    else
+    {
+#>
+                    return InvokeWithoutReturn(argsAndResults);
```

Comment:
By using a sync function that `return`s after calling the async `InvokeWithReturn`, it means the `argsAndResults` array would be returned back to the `ArrayPool` once the first `await Task.Yield()` in the async `Invoke()` is executed, which could mean other parts of the app might write something into the array while the `wasmtime_call_future_t` is still alive.

I think the code probably needs to be changed here to be an `async` function and then use `return await InvokeWithReturn()` or `await InvokeWithourReturn()`, to ensure the code from the `finally` block (which returns the array to the pool) is only executed after the `Task` is completed. What do you think?

```suggestion
            <#= delegateType #> result = async (<#
    for (int x = 0; x < parameterCount; x++)
    {
        if (x >= 1)
        {
            #>, <#
        }

        #>p<#= x.ToString(CultureInfo.InvariantCulture) #><#

    }
#>) =>
            {
                // Allocate space for both the arguments and the results.
                var argsAndResults = ArrayPool<Value>.Shared.Rent(allocCount);
                try
                {
<#
    for (int x = 0; x < parameterCount; x++)
    {
        string xStr = x.ToString(CultureInfo.InvariantCulture);
#>
                    argsAndResults[<#= xStr #>] = Value.FromValueBox(
                        convT<#= parameterCount is 1 ? "" : (x + 1).ToString(CultureInfo.InvariantCulture) #>.Box(p<#= xStr #>)
                    );
<#
    }

    if (returnTypeCount > 0)
    {
#>
                    return await InvokeWithReturn(argsAndResults, factory);
<#
    }
    else
    {
#>
                    await InvokeWithoutReturn(argsAndResults);
```

Reply:
Oops, well spotted!